daisi/src/cpps/logical/order_management/simple_order_management.cpp

```diff
@@ -0,0 +1,199 @@
+//
+// You should have received a copy of the GNU General Public License along with DAISI. If not, see
+// <https://www.gnu.org/licenses/>.
+//
+// SPDX-License-Identifier: GPL-2.0-only
+
+#include "simple_order_management.h"
+
+using namespace daisi::material_flow;
+namespace daisi::cpps::logical {
+
+template <class> inline constexpr bool always_false_v = false;
+
+SimpleOrderManagement::SimpleOrderManagement(const AmrDescription &amr_description,
+                                             const Topology &topology,
+                                             const daisi::util::Pose &pose)
+    : OrderManagement(amr_description, topology, pose),
+      current_task_(),
+      current_ordering_(),
+      current_metrics_(),
+      current_end_position_(pose.position),
+      time_now_(0) {}
+
+Metrics SimpleOrderManagement::getCurrentMetrics() const { return current_metrics_; }
+
+void SimpleOrderManagement::setCurrentTime(const daisi::util::Duration &now) {
+  if (now < time_now_) {
+    throw std::invalid_argument("new time must be later than current time");
+  }
+  time_now_ = now;
+}
+
+bool SimpleOrderManagement::hasTasks() const { return current_task_.has_value(); }
+
+Task SimpleOrderManagement::getCurrentTask() const {
+  if (!hasTasks()) {
+    throw std::logic_error("no tasks available");
+  }
+  return current_task_.value();
+}
+
+bool SimpleOrderManagement::setNextTask() {
+  if (!current_ordering_.empty()) {
+    current_task_ = current_ordering_.front();
+    current_ordering_.erase(current_ordering_.begin());
+    return true;
+  }
+  current_task_ = std::nullopt;
+  return false;
+}
+
+bool SimpleOrderManagement::canAddTask(const Task &task) {
+  SimpleOrderManagement copy(*this);
+  bool result = copy.addTask(task);
+
+  return result;
+}
+
+bool SimpleOrderManagement::addTask(const Task &task) {
+  // simply add all orders in the given order
+  const auto orders = task.getOrders();
+  if (orders.empty()) {
+    throw std::invalid_argument("Task must have at least one order");
+  }
+
+  current_ordering_.push_back(task);
+
+  // update current_metrics to match the final order of the new task
+  updateCurrentMetrics();
+
+  // find final order with an endpoint to update current_end_position_
+  std::optional<Order> final_order = std::nullopt;
+  for (auto rev_orders_it = orders.rbegin(); rev_orders_it != orders.rend(); rev_orders_it++) {
+    if (std::holds_alternative<TransportOrder>(*rev_orders_it) ||
+        std::holds_alternative<MoveOrder>(*rev_orders_it)) {
+      final_order = *rev_orders_it;
+      break;
+    }
+  }
+  if (!final_order.has_value()) {
+    throw std::logic_error("Task must contain at least one TransportOrder or MoveOrder");
+  }
+  if (std::holds_alternative<TransportOrder>(final_order.value())) {
+    TransportOrder transport_order = std::get<TransportOrder>(final_order.value());
+    current_end_position_ =
+        transport_order.getDeliveryTransportOrderStep().getLocation().getPosition();
```

Comment:
Its usually better to use get_if instead of the combination of holds_alternative and get, because we need to write out `TransportOrder` twice. 

Use for example: `if(auto transport_order = std::get_if<TransportOrder>(&final_order) {...}`. Then you can continue working on the `transport_order` pointer.

Same for `MoveOrder`.

Reply:
I agree that the use of `get_if` is more practical here. However, I found out that this part can be shortened by using `getEndLocationOfOrder`:

`auto end_location = getEndLocationOfOrder(final_order.value());
  current_end_position_ = end_location->getPosition();`